File: kotlin/bank-account/src/main/kotlin/BankAccount.kt
class BankAccount {
    private var open = true
    private var currentBalance = 0

    val balance
        get(): Int {
            check(open)
            return this.currentBalance
        }

    @Synchronized
    fun adjustBalance(amount: Int) {
        check(open)
        currentBalance += amount
    }

    @Synchronized
    fun close() {
        open = false
    }
}

File: clojure/bank-account/src/bank_account.clj
(ns bank-account)

(defrecord BankAccount [balance open?])

(defn open-account []
  (atom (BankAccount. 0 true)))

(defn close-account [bank-account]
  (swap! bank-account assoc :open? false))

(defn get-balance [bank-account]
  (when (:open? @bank-account)
    (:balance @bank-account)))

(defn update-balance [bank-account adjustment]
  (locking bank-account
    (swap! bank-account assoc :balance (+ (get-balance bank-account) adjustment))))
